### PR TITLE
Fix cgroup config

### LINKED ISSUE #842: Incorrect cgroup when using systemd cgroup driver
Steps to reproduce:

```
# verify systemd cgroup driver
]$ cat /etc/crio/crio.conf | grep cgroup_manager
# cgroup_manager is the cgroup management implementation to be used
cgroup_manager = "systemd"
# spin up kubernetes with matching driver
$ cd kubernetes
$ export CGROUP_DRIVER=systemd
$ export KUBE_ENABLE_CLUSTER_DASHBOARD=true
$ export KUBE_ENABLE_CLUSTER_DNS=true
$ hack/local-up-cluster.sh
..
$ kubectl get pods --all-namespaces
# should see the dns and dashboard pods running
```

Verify the cgroup tree:

kubelet creates the following pod cgroup:

```
/sys/fs/cgroup/cpu/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod12723d69_9268_11e7_84d0_c85b76cda386.slice
```

note: kubepods-burstable-pod12723d69_9268_11e7_84d0_c85b76cda386.slice

there is nothing in the cgroupfs for the container, CRI-O is creating them here:

```
/sys/fs/cgroup/cpu/kubepods.slice.slice/kubepods.slice-kubepods_burstable.slice.slice/kubepods.slice-kubepods_burstable.slice-kubepods_burstable_pod12723d69_9268_11e7_84d0_c85b76cda386.slice.slice
```

note: the double encoding everywhere in the hierarchy

i think crio when running with cgroup manager = systemd should error when given invalid cgroups, but it should just convert from cgroupfs to systemd naming doing something similar to here:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cgroup_manager_linux.go#L109

### PR DESCRIPTION
**Pull request by mrunalp**
Fixes #842

We ensure that cri-o cgroup manager is compatible with the cgroup parent sent by k8s and do the right conversions to pass the cgroups to runc.

### CODE REVIEW

File: server/sandbox_run.go [hunk 1]
```diff
@@ -331,13 +331,19 @@ func (s *Server) RunPodSandbox(ctx context.Context, req *pb.RunPodSandboxRequest
 	cgroupParent := req.GetConfig().GetLinux().CgroupParent
 	if cgroupParent != "" {
 		if s.config.CgroupManager == oci.SystemdCgroupsManager {
-			cgPath, err := convertCgroupNameToSystemd(cgroupParent, false)
+			if !strings.HasSuffix(path.Base(cgroupParent), ".slice") {
```
Comment (runcom) [inline server/sandbox_run.go]: `if len(config.CgroupParent) <= 6 || !strings.HasSuffix(config.CgroupParent, ".slice")`

**Comment by runcom** [comment]
Seems ok, but dockershim in kubelet is doing thing differently https://github.com/kubernetes/kubernetes/blob/4d409a4d9e9772193cb24944bd76b90aaca01473/pkg/kubelet/dockershim/docker_service.go#L393 and it's always converting to systemd, which might mean it has the same bug

**Comment by runcom** [comment]
I'm testing what happens with Docker

**Comment by runcom** [comment]
So, spent some time analyzing the whole stuff here https://github.com/kubernetes-incubator/cri-o/issues/842#issuecomment-327428272

The result is, @mrunalp could you cherry-pick these 2 commits from #814:

- https://github.com/kubernetes-incubator/cri-o/pull/814/commits/998a649cd4f8519e3a0b01d4604407b8996c42fe
- https://github.com/kubernetes-incubator/cri-o/pull/814/commits/f3565d327199bcd44f0315e5c7b694e120367188

These 2 commits + your PR here will fix all cgroup issues.

File: server/sandbox_run.go (hunk 1 above)
Comment (TomSweeneyRedHat) [reply server/sandbox_run.go]: Would it be at all useful to have different error messages for these conditions?  I think with this suggestion a small slice would trigger the below error saying we'd not received a slice when we in fact had.  Or maybe I've not had enough (any) tea yet this morning.

Comment (runcom) [reply server/sandbox_run.go]: the check I'm suggesting is to handler `cgroupParent = ".slice"` - docker handles that as well.

**Comment by mrunalp** [comment]
Updated.

**Comment by runcom** [comment]
Looks good but we may want to change the sandbox_run cgpath code to look similar to this https://github.com/kubernetes-incubator/cri-o/commit/998a649cd4f8519e3a0b01d4604407b8996c42fe to handle defaults (w/o cgroup parent)

**Comment by runcom** [comment]
(note to self: we need to re-work https://github.com/kubernetes-incubator/cri-o/pull/814/commits/998a649cd4f8519e3a0b01d4604407b8996c42fe to use what in this PR in order to handle cases when cgroupparent == "", can't happen with kube though but still..)

**Comment by mrunalp** [comment]
Updated.

**Comment by runcom** [comment]
LGTM

File: server/sandbox_run.go [hunk 2]
```diff
@@ -572,40 +579,14 @@ func setupShm(podSandboxRunDir, mountLabel string) (shmPath string, err error) {
...
-		if err != nil {
-			return "", fmt.Errorf("error adapting cgroup name, input: %v, err: %v", name, err)
-		}
-	}
-	return result, nil
+// convertCgroupFsNameToSystemd converts an expanded cgroupfs name to its systemd name.
+// For example, it will convert test.slice/test-a.slice/test-a-b.slice to become test-a-b.slice
+// NOTE: this is public right now to allow its usage in dockermanager and dockershim, ideally both those
+// code areas could use something from libcontainer if we get this style function upstream.
+func convertCgroupFsNameToSystemd(cgroupfsName string) (string, error) {
```
Comment (runcom) [inline server/sandbox_run.go]: This could be used directly from kube pkg but let's defer this to a follow up

Comment (mrunalp) [reply server/sandbox_run.go]: Reason is it was pulling in tons of vendors for this simple function. We could revisit this later.

Comment (runcom) [reply server/sandbox_run.go]: Definitely, I didn't mean to look at this now either

**Comment by runcom** [comment]
[retest]

**Comment by mrunalp** [comment]
@runcom This is green :)

**Pull request merged**